File: FitoGrano/CS/WinForm/frBooks_ResearchObjects.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using System.Linq;
using System.Drawing;
using System.Diagnostics;
using Microsoft.VisualBasic;
using System.Xml.Linq;
using System.Windows.Forms;
// End of VB project level imports

using System.Data;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Navigation;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.ButtonsPanelControl;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using FitoGrano;

namespace FitoGrano
{
	
	public partial class frBooks_ResearchObjects
	{
		public frBooks_ResearchObjects()
		{
			InitializeComponent();
			
			//Added to support default instance behavour in C#
			if (defaultInstance == null)
				defaultInstance = this;
		}
		
#region Default Instance
		
		private static frBooks_ResearchObjects defaultInstance;
		
		/// <summary>
		/// Added by the VB.Net to C# Converter to support default instance behavour in C#
		/// </summary>
		public static frBooks_ResearchObjects Default
		{
			get
			{
				if (defaultInstance == null)
				{
					defaultInstance = new frBooks_ResearchObjects();
					defaultInstance.FormClosed += new System.Windows.Forms.FormClosedEventHandler(defaultInstance_FormClosed);
				}
				
				return defaultInstance;
			}
			set
			{
				defaultInstance = value;
			}
		}
		
		static void defaultInstance_FormClosed(object sender, System.Windows.Forms.FormClosedEventArgs e)
		{
			defaultInstance = null;
		}
		
#endregion
		string PreferenceForms; // Ветка в реестре для хранения настроек формы
		string SelectedObjectType; // активный тип объекта
		internal string NewId_Object; // Ид нового объекта
		// закрытие формы
		public void frBooks_FormClosing(object sender, FormClosingEventArgs e)
		{
			clsMyUserFunction.RemoveTableDataSet(this);
			clsMyUserFunction.SaveViewForm(this);
			mdRegistrySetting.RegistryWrite(PreferenceForms, "SplitterPosition", System.Convert.ToString(this.SplitContainerControl1.SplitterPosition));
			mdRegistrySetting.RegistryWrite(PreferenceForms, "gvObjects.ShowFilterRow", this.gvObjects.OptionsView.ShowAutoFilterRow.ToString());
			mdRegistrySetting.RegistryWrite(PreferenceForms, "gvObjects.ShowFindPanel", this.gvObjects.OptionsFind.AlwaysVisible.ToString());
			mdRegistrySetting.RegistryWrite(PreferenceForms, "SelectedObject", this.Accord.SelectedElement.Name);
		}
		// загрузка формы
		public void frBooks_ResearchObjects_Load(object sender, EventArgs e)
		{
			this.acRoot.HeaderVisible = false;
			// загрузка объектов
			Designer_ObjectsTypes();
			// настройка элементов
			PreferenceForms = mdVariables.pref_UserSettings + this.Name + "\\";
			clsMyUserFunction.LoadViewForm(this);
			this.SplitContainerControl1.SplitterPosition = System.Convert.ToInt32(mdRegistrySetting.RegistryRead(PreferenceForms, "SplitterPosition", System.Convert.ToString(180)));
			this.gcnObjects.CustomHeaderButtons[4].Properties.Checked = System.Convert.ToBoolean(mdRegistrySetting.RegistryRead(PreferenceForms, "gvObjects.ShowFindPanel", System.Convert.ToString(false)));
			this.gcnObjects.CustomHeaderButtons[5].Properties.Checked = System.Convert.ToBoolean(mdRegistrySetting.RegistryRead(PreferenceForms, "gvObjects.ShowFilterRow", System.Convert.ToString(false)));
			this.Accord.SelectedElement = this.acRoot.Elements[System.Convert.ToString(mdRegistrySetting.RegistryRead(PreferenceForms, "SelectedObject", this.acRoot.Elements[0].Name))];
			this.lbDB_name.Caption = mdVariables.pref_DataBaseName;
			this.lbServerName.Caption = mdVariables.pref_ServerIP;
		}
		// выбор типа объекта
		public void Accord_SelectedElementChanged(object sender, SelectedElementChangedEventArgs e)
		{
			SelectedObjectType = e.Element.Name;
			Designer_Objects();
		}
		// группа кнопок объектов
		public void gcnObjects_CustomButtonClick(object sender, DevExpress.XtraBars.Docking2010.BaseButtonEventArgs e)
		{
			bool btnAddEdit = false;
			GroupBoxButton btn = e.Button;
			switch (btn.Caption)
			{
				case "btnAddObject":
					frAddEditObjects.Default.AddEdit = 2;
					btnAddEdit = true;
					break;
				case "btnEditObject":
					if (this.gvObjects.RowCount != 0)
					{
						frAddEditObjects.Default.AddEdit = 3;
						btnAddEdit = true;
					}
					break;
				case "btnDeleteObject":
					if (this.gvObjects.RowCount != 0)
					{
						btnDeleteObject();
					}
					break;
				case "btnClearSort_Object":
					this.gvObjects.ClearSorting();
					break;
			}
			// если добавление или редактирование
			if (btnAddEdit)
			{
				if (frAddEditObjects.Default.ShowDialog() == DialogResult.OK)
				{
					Designer_Objects();
					Designer_ObjectsTypes();
				}
			}
		}
		// группа чеков объектов
		public void gcnMethods_CustomButtonChecked(object sender, DevExpress.XtraBars.Docking2010.BaseButtonEventArgs e)
		{
			GroupBoxButton btn = e.Button;
			switch (btn.Caption)
			{
				case "btnSearch_Object":
					this.gvObjects.OptionsFind.AlwaysVisible = btn.Checked;
					// при отключении очищаем фильтрацию
					if (btn.Checked == false)
					{
						this.gvObjects.FindFilterText = null;
					}
					break;
				case "btnAutoFilter_Object":
					this.gvObjects.OptionsView.ShowAutoFilterRow = btn.Checked;
					// при отключении очищаем фильтрацию
					if (btn.Checked == false)
					{
						this.gvObjects.ClearColumnsFilter();
					}
					break;
			}
		}
		// настройки сетки типов объектов
		private void Designer_ObjectsTypes()
		{
			// загрузка объектов
			mdQuery.SelectQueryData("Books_ObjectsTypes." + this.Name, "SELECT * FROM vBooks_ObjectsTypes", "GetObjectsTypes");
			this.acRoot.Elements.Clear();
			foreach (DataRow r in mdQuery.iDataSet.Tables["Books_ObjectsTypes." + this.Name].Rows)
			{
				AccordionControlElement el = new AccordionControlElement();
				el.Name = "btn" + System.Convert.ToString(r["Name"]);
				el.Text = "<image=#" + System.Convert.ToString(r["Icon"]) + ">   " + System.Convert.ToString(r["NameCountHtml"]);
				el.Tag = r["Id"];
				el.Style = ElementStyle.Item;
				el.Height = 40;
				this.acRoot.Elements.Add(el);
			}
			this.Accord.SelectedElement = this.acRoot.Elements[SelectedObjectType];
		}
		// настройки сетки объектов
		private void Designer_Objects()
		{
			mdVariables.EventChangedControl = false;
			mdQuery.SelectQueryData("Books_Objects." + this.Name,
				"EXEC Books_ObjectsFunc @ObjectTypeId = " + System.Convert.ToString(this.Accord.SelectedElement.Tag) +, System.Convert.ToString(
				@Function == 1));
			if (ReferenceEquals(this.gcObjects.DataSource, null))
			{
				this.gcObjects.DataSource = mdQuery.iDataSet.Tables["Books_Objects." + this.Name];
				mdHidenAllColumns.HidenAllColumns_Grid(this.gvObjects, mdQuery.iDataSet.Tables["Books_Objects." + this.Name]);
				this.gvObjects.Columns["NameLatin"].Visible = true;
				this.gvObjects.Columns["NameLatin"].Caption = "Наименование(латин)";
				this.gvObjects.Columns["Name"].Visible = true;
				this.gvObjects.Columns["Name"].Caption = "Наименование (русс)";
				this.gvObjects.PreviewFieldName = "MethodsString";
				this.gvObjects.BestFitColumns();
			}
			mdVariables.EventChangedControl = true;
			// активная строка из памяти
			this.gvObjects.FocusedRowHandle = this.gvObjects.LocateByValue("Id", System.Convert.ToInt32(NewId_Object));
		}
		// удалить или редактировать объект клавиотурой
		public void gvObjects_KeyDown(dynamic sender, KeyEventArgs e)
		{
			if (sender.FocusedRowHandle < 0)
			{
				return;
			}
			if (e.KeyCode == Keys.Delete)
			{
				btnDelObject_ItemClick(e, null);
			}
			if (e.KeyCode == Keys.Enter)
			{
				btnEditObject_ItemClick(e, null);
			}
		}
		// удаление объекта
		private void btnDeleteObject()
		{
			int RowNumber = this.gvObjects.FocusedRowHandle;
			if (XtraMessageBox.Show("Объект исследования с именем <b>" + this.gvObjects.GetFocusedRowCellDisplayText("Name") + "</b> будет удален!" + "\n" +
					"Вы согласны?", Application.ProductName, MessageBoxButtons.YesNo,
					MessageBoxIcon.Question, DevExpress.Utils.DefaultBoolean.True) == DialogResult.Yes)
					{
					mdQuery.ExecuteScalar(
						"EXEC Books_ObjectsFunc " +
						"@ObjectId = " + this.gvObjects.GetFocusedRowCellDisplayText("Id") + ", " +
						"@Function = 4",
						"Books_ClassificationsProducts.Function = 4"
						);
				Designer_Objects();
				Designer_ObjectsTypes();
				this.gvObjects.FocusedRowHandle = RowNumber - 1;
			}
		}
		// вызов контекстного меню на объектах
		public void gvObjects_MouseDown(object sender, MouseEventArgs e)
		{
			GridView view = (GridView) sender;
			Point pt = view.GridControl.PointToClient(Control.MousePosition);
			GridHitInfo info = view.CalcHitInfo(pt);
			if (e.Button == MouseButtons.Right)
			{
				this.btnEditObject.Enabled = info.InRow;
				this.btnDelObject.Enabled = info.InRow;
				this.pmObjects.ShowPopup(MousePosition);
			}
			if (info.InRow & e.Button == MouseButtons.Left & e.Clicks == 2)
			{
				btnEditObject_ItemClick(e, null);
			}
		}
		// фокус на объектах
		public void gvObjects_FocusedRowChanged(dynamic sender, DevExpress.XtraGrid.Views.Base.FocusedRowChangedEventArgs e)
		{
			if (e.FocusedRowHandle < 0)
			{
				return;
			}
			DataRow iRow = sender.GetFocusedDataRow;
			if (mdVariables.EventChangedControl)
			{
				NewId_Object = iRow["Id"].ToString();
			}
		}
		
#region Контекстное меню на объектах
		// добавить объект
		public void btnAddObject_ItemClick(object sender, ItemClickEventArgs e)
		{
			gcnObjects_CustomButtonClick(this.gcnObjects, new DevExpress.XtraBars.Docking2010.BaseButtonEventArgs(this.gcnObjects.CustomHeaderButtons[0]));
		}
		// редактировать объект
		public void btnEditObject_ItemClick(object sender, ItemClickEventArgs e)
		{
			gcnObjects_CustomButtonClick(this.gcnObjects, new DevExpress.XtraBars.Docking2010.BaseButtonEventArgs(this.gcnObjects.CustomHeaderButtons[1]));
		}
		// удалить объект
		public void btnDelObject_ItemClick(object sender, ItemClickEventArgs e)
		{
			gcnObjects_CustomButtonClick(this.gcnObjects, new DevExpress.XtraBars.Docking2010.BaseButtonEventArgs(this.gcnObjects.CustomHeaderButtons[2]));
		}
#endregion
		
		// кнопка Обновить
		public void btnUpdate_ItemClick(object sender, ItemClickEventArgs e)
		{
			Designer_Objects();
		}
		// проверка полей по таймеру
		public void tmChecked_Tick(object sender, EventArgs e)
		{
			int ObjectsCount = this.gvObjects.RowCount;
			this.gcnObjects.CustomHeaderButtons[0].Properties.Enabled = this.acRoot.Elements.Count != 0;
			this.btnAddObject.Enabled = this.acRoot.Elements.Count != 0;
			this.gcnObjects.CustomHeaderButtons[1].Properties.Enabled = ObjectsCount != 0;
			this.btnEditObject.Enabled = ObjectsCount != 0;
			this.gcnObjects.CustomHeaderButtons[2].Properties.Enabled = ObjectsCount != 0;
			this.btnDelObject.Enabled = ObjectsCount != 0;
			// подсчет столбцов с примененной сортировкой
			int ColCount = 0;
			foreach (GridColumn c in this.gvObjects.SortedColumns)
			{
				ColCount++;
			}
			this.gcnObjects.CustomHeaderButtons[6].Properties.Enabled = ColCount != 0;
		}
	}
}

File: FitoGrano/WinForm/frBooks_ResearchObjects.vb
Imports System.Data
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Navigation
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.ButtonsPanelControl
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Public Class frBooks_ResearchObjects
    Dim PreferenceForms As String       ' Ветка в реестре для хранения настроек формы 
    Dim SelectedObjectType As String        ' активный тип объекта
    Friend NewId_Object As String       ' Ид нового объекта 
    ' закрытие формы
    Private Sub frBooks_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        clsMyUserFunction.RemoveTableDataSet(Me)
        clsMyUserFunction.SaveViewForm(Me)
        RegistryWrite(PreferenceForms, "SplitterPosition", Me.SplitContainerControl1.SplitterPosition)
        RegistryWrite(PreferenceForms, "gvObjects.ShowFilterRow", Me.gvObjects.OptionsView.ShowAutoFilterRow.ToString)
        RegistryWrite(PreferenceForms, "gvObjects.ShowFindPanel", Me.gvObjects.OptionsFind.AlwaysVisible.ToString)
        RegistryWrite(PreferenceForms, "SelectedObject", Me.Accord.SelectedElement.Name)
    End Sub
    ' загрузка формы
    Private Sub frBooks_ResearchObjects_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Me.acRoot.HeaderVisible = False
        ' загрузка объектов
        Designer_ObjectsTypes()
        ' настройка элементов
        PreferenceForms = pref_UserSettings & Me.Name & "\"
        clsMyUserFunction.LoadViewForm(Me)
        Me.SplitContainerControl1.SplitterPosition = RegistryRead(PreferenceForms, "SplitterPosition", 180)
        Me.gcnObjects.CustomHeaderButtons.Item(4).Properties.Checked = RegistryRead(PreferenceForms, "gvObjects.ShowFindPanel", False)
        Me.gcnObjects.CustomHeaderButtons.Item(5).Properties.Checked = RegistryRead(PreferenceForms, "gvObjects.ShowFilterRow", False)
        Me.Accord.SelectedElement = Me.acRoot.Elements.Item(RegistryRead(PreferenceForms, "SelectedObject", Me.acRoot.Elements.Item(0).Name))
        Me.lbDB_name.Caption = pref_DataBaseName
        Me.lbServerName.Caption = pref_ServerIP
    End Sub
    ' выбор типа объекта
    Private Sub Accord_SelectedElementChanged(sender As Object, e As SelectedElementChangedEventArgs) Handles Accord.SelectedElementChanged
        SelectedObjectType = e.Element.Name
        Designer_Objects()
    End Sub
    ' группа кнопок объектов
    Private Sub gcnObjects_CustomButtonClick(sender As Object, e As Docking2010.BaseButtonEventArgs) Handles gcnObjects.CustomButtonClick
        Dim btnAddEdit As Boolean = False
        Dim btn As GroupBoxButton = e.Button
        Select Case btn.Caption
            Case "btnAddObject" : frAddEditObjects.AddEdit = 2 : btnAddEdit = True
            Case "btnEditObject" : If Me.gvObjects.RowCount <> 0 Then frAddEditObjects.AddEdit = 3 : btnAddEdit = True
            Case "btnDeleteObject" : If Me.gvObjects.RowCount <> 0 Then btnDeleteObject()
            Case "btnClearSort_Object" : Me.gvObjects.ClearSorting()
        End Select
        ' если добавление или редактирование
        If btnAddEdit Then
            If frAddEditObjects.ShowDialog() = DialogResult.OK Then
                Designer_Objects()
                Designer_ObjectsTypes()
            End If
        End If
    End Sub
    ' группа чеков объектов
    Private Sub gcnMethods_CustomButtonChecked(sender As Object, e As Docking2010.BaseButtonEventArgs) Handles gcnObjects.CustomButtonChecked, gcnObjects.CustomButtonUnchecked
        Dim btn As GroupBoxButton = e.Button
        Select Case btn.Caption
            Case "btnSearch_Object"
                Me.gvObjects.OptionsFind.AlwaysVisible = btn.Checked
                ' при отключении очищаем фильтрацию
                If btn.Checked = False Then Me.gvObjects.FindFilterText = Nothing
            Case "btnAutoFilter_Object"
                Me.gvObjects.OptionsView.ShowAutoFilterRow = btn.Checked
                ' при отключении очищаем фильтрацию
                If btn.Checked = False Then Me.gvObjects.ClearColumnsFilter()
        End Select
    End Sub
    ' настройки сетки типов объектов
    Private Sub Designer_ObjectsTypes()
        ' загрузка объектов
        SelectQueryData("Books_ObjectsTypes." & Me.Name, "SELECT * FROM vBooks_ObjectsTypes", "GetObjectsTypes")
        Me.acRoot.Elements.Clear()
        For Each r As DataRow In iDataSet.Tables("Books_ObjectsTypes." & Me.Name).Rows
            Dim el As New AccordionControlElement
            el.Name = "btn" & r("Name")
            el.Text = "<image=#" & r("Icon") & ">   " & r("NameCountHtml")
            el.Tag = r("Id")
            el.Style = ElementStyle.Item
            el.Height = 40
            Me.acRoot.Elements.Add(el)
        Next
        Me.Accord.SelectedElement = Me.acRoot.Elements.Item(SelectedObjectType)
    End Sub
    ' настройки сетки объектов
    Private Sub Designer_Objects()
        EventChangedControl = False
        SelectQueryData("Books_Objects." & Me.Name,
                        "EXEC Books_ObjectsFunc @ObjectTypeId = " & Me.Accord.SelectedElement.Tag & ", 
                        @Function = 1")
        If Me.gcObjects.DataSource Is Nothing Then
            Me.gcObjects.DataSource = iDataSet.Tables("Books_Objects." & Me.Name)
            mdHidenAllColumns.HidenAllColumns_Grid(Me.gvObjects, iDataSet.Tables("Books_Objects." & Me.Name))
            Me.gvObjects.Columns("NameLatin").Visible = True
            Me.gvObjects.Columns("NameLatin").Caption = "Наименование(латин)"
            Me.gvObjects.Columns("Name").Visible = True
            Me.gvObjects.Columns("Name").Caption = "Наименование (русс)"
            Me.gvObjects.PreviewFieldName = "MethodsString"
            Me.gvObjects.BestFitColumns()
        End If
        EventChangedControl = True
        ' активная строка из памяти
        Me.gvObjects.FocusedRowHandle = Me.gvObjects.LocateByValue("Id", CType(NewId_Object, Integer))
    End Sub
    ' удалить или редактировать объект клавиотурой
    Private Sub gvObjects_KeyDown(sender As Object, e As KeyEventArgs) Handles gvObjects.KeyDown
        If sender.FocusedRowHandle < 0 Then Exit Sub
        If e.KeyCode = Keys.Delete Then
            btnDelObject_ItemClick(e, Nothing)
        End If
        If e.KeyCode = Keys.Enter Then
            btnEditObject_ItemClick(e, Nothing)
        End If
    End Sub
    ' удаление объекта
    Private Sub btnDeleteObject()
        Dim RowNumber As Integer = Me.gvObjects.FocusedRowHandle
        If XtraMessageBox.Show("Объект исследования с именем <b>" & Me.gvObjects.GetFocusedRowCellDisplayText("Name") & "</b> будет удален!" & Chr(10) &
                               "Вы согласны?", Application.ProductName, MessageBoxButtons.YesNo,
                               MessageBoxIcon.Question, DevExpress.Utils.DefaultBoolean.True) = DialogResult.Yes Then
            ExecuteScalar(
                          "EXEC Books_ObjectsFunc " &
                               "@ObjectId = " & Me.gvObjects.GetFocusedRowCellDisplayText("Id") & ", " &
                               "@Function = 4",
                          "Books_ClassificationsProducts.Function = 4"
                          )
            Designer_Objects()
            Designer_ObjectsTypes()
            Me.gvObjects.FocusedRowHandle = RowNumber - 1
        End If
    End Sub
    ' вызов контекстного меню на объектах
    Private Sub gvObjects_MouseDown(sender As Object, e As MouseEventArgs) Handles gvObjects.MouseDown
        Dim view As GridView = CType(sender, GridView)
        Dim pt As Point = view.GridControl.PointToClient(Control.MousePosition)
        Dim info As GridHitInfo = view.CalcHitInfo(pt)
        If e.Button = MouseButtons.Right Then
            Me.btnEditObject.Enabled = info.InRow
            Me.btnDelObject.Enabled = info.InRow
            Me.pmObjects.ShowPopup(MousePosition)
        End If
        If info.InRow And e.Button = MouseButtons.Left And e.Clicks = 2 Then
            btnEditObject_ItemClick(e, Nothing)
        End If
    End Sub
    ' фокус на объектах
    Private Sub gvObjects_FocusedRowChanged(sender As Object, e As DevExpress.XtraGrid.Views.Base.FocusedRowChangedEventArgs) Handles gvObjects.FocusedRowChanged
        If e.FocusedRowHandle < 0 Then Exit Sub
        Dim iRow As DataRow = sender.GetFocusedDataRow
        If EventChangedControl Then NewId_Object = iRow("Id").ToString
    End Sub

#Region "Контекстное меню на объектах"
    ' добавить объект
    Private Sub btnAddObject_ItemClick(sender As Object, e As ItemClickEventArgs) Handles btnAddObject.ItemClick
        gcnObjects_CustomButtonClick(Me.gcnObjects, New Docking2010.BaseButtonEventArgs(Me.gcnObjects.CustomHeaderButtons.Item(0)))
    End Sub
    ' редактировать объект
    Private Sub btnEditObject_ItemClick(sender As Object, e As ItemClickEventArgs) Handles btnEditObject.ItemClick
        gcnObjects_CustomButtonClick(Me.gcnObjects, New Docking2010.BaseButtonEventArgs(Me.gcnObjects.CustomHeaderButtons.Item(1)))
    End Sub
    ' удалить объект
    Private Sub btnDelObject_ItemClick(sender As Object, e As ItemClickEventArgs) Handles btnDelObject.ItemClick
        gcnObjects_CustomButtonClick(Me.gcnObjects, New Docking2010.BaseButtonEventArgs(Me.gcnObjects.CustomHeaderButtons.Item(2)))
    End Sub
#End Region

    ' кнопка Обновить
    Private Sub btnUpdate_ItemClick(sender As Object, e As ItemClickEventArgs) Handles btnUpdate.ItemClick
        Designer_Objects()
    End Sub
    ' проверка полей по таймеру
    Private Sub tmChecked_Tick(sender As Object, e As EventArgs) Handles tmChecked.Tick
        Dim ObjectsCount As Integer = Me.gvObjects.RowCount
        Me.gcnObjects.CustomHeaderButtons.Item(0).Properties.Enabled = Me.acRoot.Elements.Count <> 0
        Me.btnAddObject.Enabled = Me.acRoot.Elements.Count <> 0
        Me.gcnObjects.CustomHeaderButtons.Item(1).Properties.Enabled = ObjectsCount <> 0
        Me.btnEditObject.Enabled = ObjectsCount <> 0
        Me.gcnObjects.CustomHeaderButtons.Item(2).Properties.Enabled = ObjectsCount <> 0
        Me.btnDelObject.Enabled = ObjectsCount <> 0
        ' подсчет столбцов с примененной сортировкой
        Dim ColCount As Integer = 0
        For Each c As GridColumn In Me.gvObjects.SortedColumns
            ColCount += 1
        Next
        Me.gcnObjects.CustomHeaderButtons.Item(6).Properties.Enabled = (ColCount <> 0)
    End Sub
End Class